connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java

```diff
@@ -1592,6 +1731,28 @@ public void onSessionKeyUpdate(SessionKey sessionKey) {
                 }
             }
         }
+
+        @Override
+        public void onRestartRequest(RestartRequest request) {
+            log.info("Received and enqueuing {}", request);
+
+            synchronized (DistributedHerder.this) {
+                String connectorName = request.connectorName();
+                //preserve the highest impact request
+                if (pendingRestartRequests.containsKey(connectorName)) {
+                    RestartRequest existingRequest = pendingRestartRequests.get(connectorName);
+                    if (request.compareTo(existingRequest) > 0) {
+                        log.debug("Overwriting existing {} and enqueuing the higher impact {}", existingRequest, request);
+                        pendingRestartRequests.put(connectorName, request);
+                    } else {
+                        log.debug("Preserving existing higher impact {} and ignoring incoming {}", existingRequest, request);
+                    }
+                } else {
+                    pendingRestartRequests.put(connectorName, request);
+                }
```

Comment:
If we were to change `pendingRestartRequests` to a `ConcurrentMap`, then we could use the `compute(...)` functionality that actually would work even if we removed the synchronization. I wonder if this is a bit more readable anyway. WDYT?
```
                pendingRestartRequests.compute(connectorName, (k, existing) -> {
                    if (existing == null || request.compareTo(existing) > 0) {
                        log.debug("Overwriting existing {} and enqueuing the higher impact {}", existing, request);
                        return request;
                    }
                    log.debug("Preserving existing higher impact {} and ignoring incoming {}", existing, request);
                    return existing;
                });
```

Reply:
turns out a normal hashmap also has compute so I just used it.